packages/imperative/src/doc/IApimlSvcAttrs.ts

```diff
@@ -0,0 +1,71 @@
+/*
+* This program and the accompanying materials are made available under the terms of the
+* Eclipse Public License v2.0 which accompanies this distribution, and is available at
+* https://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Copyright Contributors to the Zowe Project.
+*
+*/
+
+/**
+ * When a command group or plugin is used with the API Mediation Layer,
+ * the connection properties for communicating to the service associated
+ * with the command group can be retrieved from APIML. These attributes
+ * are used to identify the desired REST service within APIML.
+ *
+ * Information retrieved from APIML using these attributes is then used
+ * to automatically create connection profile properties into a Zowe
+ * team configuration file.
+ *
+ * Note that the "ibm.zosmf" service apiId is baked into the Zowe CLI.
+ * Therefore, this optional interface need not be specified by a plugin
+ * that requires only a connection to z/OSMF.
+ */
+export interface IApimlSvcAttrs {
+    /**
+     * The identifier of the REST API to which this command group communicates.
+     * The API ID is defined by the REST service programmer. An example is
+     * "zowe.apiml.apicatalog".
+     * @type {string}
+     * @memberof IApimlSvcAttrs
+     */
+    apiId: string;
+
+    /**
+     * The portion of the URL that directs a REST request to the correct path
+     * for the desired web service. When multiple versions of the same apiId
+     * are hosted on an APIML instance, this attribute identifies the desired
+     * version of the REST API. This attribute is defined by the REST service
+     * programmer. An example is "api/v1". If no gatewayUrl is supplied, but
+     * multiple versions of the REST API exists in APIML, the first instance
+     * of the apiId returned by APIML will be used. You might choose to not
+     * specify a gatewayURL, if your client app supports all available versions
+     * of the associated REST service.
+     * @type {string}
+     * @memberof IApimlSvcAttrs
+     */
+    gatewayUrl?: string;
```

Comment:
a topic that comes up time-to-time is what goes in imperative versus zowe cli.  since imperative is a generic cli infrastructure, should this be in the zowe cli code base?

Reply:
Yes. The team discussed this very question in a standup meeting. I believe you were on vacation at the time. This was our rationale:

- We want a new structure (IApimlSvcAttrs) to enable plugins to tie themselves to an APIML service. We can then ask APIML for the configuration data for the plugin to connect to that service.
    - IImperativeConfig is what each plugin uses to define itself and its options.
    - Thus, we want the new IApimlSvcAttrs structure to be added as a property to IImperativeConfig.
    - IImperativeConfig could not easily import IApimlSvcAttrs from zowe-cli.
- If IApimlSvcAttrs was instead added to Zowe-cli, then zowe-cli would have to extend IImperativeConfig and then add the IApimlSvcAttrs to that new extended interface.
    - All plugins would have to be modified to define themselves using the new extended Zowe-CLI interface instead of IImperativeConfig. That seemed like a big change for every plugin.
- Even though IApimlSvcAttrs is not complely generic CLI functionality (ie, non-mainframe, and non-APIML), it is optional. It is also only an interface, and no logic in imperative performs programmatic actions on the properties of IApimlSvcAttrs. We decided that these limited charactistics of IApimlSvcAttrs made it acceptable for inclusion in Imperative.

Reply:
im sorry you had to rehash that, but thank you for the explanation @gejohnston